### PR TITLE
Add method to get properties with a prefix

### LINKED ISSUE #1627: Add method to get properties with prefix to PluginEnvironment.Configuration
[PluginEnvironment.Configuration](https://github.com/apache/accumulo/blob/db9c2b975f2db1b3903402dea69e32a014d122da/core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java#L39) is a stable interface that allows plugins to access Accumulo's configuration.  Sometimes plugins need all properties with a certain prefix which currently requires iterating over all properties.  Accumulo's internal configuration class efficiently supports this in some cases.  It would be nice to add a method that supports efficiently getting props with a prefix to PluginEnvironment.Configuration.

Noticed this while working on the TooManyDeletesSelector as part of #1605 which needs to get all summarizer properties.

### PR DESCRIPTION
**Pull request by jkosh44**
Closes #1627

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java
Comment (jkosh44) [inline core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java:64]: I know this should probably have a `@since` tag, I'm just not sure what version to put. Please let me know and I'll add it.

**Comment by jkosh44** [comment]
One open question is should I be filtering out sensitive properties?

File: core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java:64]: 2.1.0 would be the version we would add this kind of thing to.

Comment (jkosh44) [reply core/src/main/java/org/apache/accumulo/core/client/PluginEnvironment.java:65]: Thanks, added

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:1247]: This method isn't necessary, since this loop can be avoided.
```suggestion
```

File: server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java:85]: This check for valid prefix property can be avoided, by checking the returned object from `Property.getPropertyByKey(prefix)`:

```suggestion
      Property propertyPrefix = Property.getPropertyByKey(prefix);
      if (propertyPrefix != null && propertyPrefix.getType() == PropertyType.PREFIX) {
```

File: server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java:95]: This can be more succinctly written with streams:
```suggestion
        return StreamSupport.stream(acfg.spliterator(), false)
            .filter(prop -> prop.getKey().startsWith(prefix))
            .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
```

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Overall, this looks good to me. I made some suggestions for improvements, and I think this could use a unit test, but otherwise, looks okay.

Are you able to include a unit test for this?

**Comment by ctubbsii** [comment]
> One open question is should I be filtering out sensitive properties?

This is for pluggable components, vetted by an admin, and placed on a server's class path. So, it has access to everything anyway. I don't think there'd be any point.

Also, did you see the test failure? I'm not sure if that's related to these changes, but might be worth looking into.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java
Comment (jkosh44) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Removed

File: server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java
Comment (jkosh44) [reply server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java]: Fixed

File: server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java
Comment (jkosh44) [reply server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java]: Fixed

**Comment by jkosh44** [comment]
> Overall, this looks good to me. I made some suggestions for improvements, and I think this could use a unit test, but otherwise, looks okay.
> 
> Are you able to include a unit test for this?

@ctubbsii I've made the changes you suggested (and I'm adding a unit test now). One scenario I was thinking about is if the provided prefix is a subset of a recognized prefix.
For example, let's say we are given the prefix "a.b.c" and "a.b" is a recognized prefix. Then it may be more efficient to check for this scenario and only iterate through "a.b"'s properties instead of iterating through all the properties. What do you think (if that made sense)?

**Comment by jkosh44** [comment]
> > One open question is should I be filtering out sensitive properties?
> 
> This is for pluggable components, vetted by an admin, and placed on a server's class path. So, it has access to everything anyway. I don't think there'd be any point.
> 
> Also, did you see the test failure? I'm not sure if that's related to these changes, but might be worth looking into.

I did not see the test failure, did it happen during the checks? If so can you send a link? I thought all the checks passed.

**Comment by ctubbsii** [comment]
> I did not see the test failure, did it happen during the checks? If so can you send a link? I thought all the checks passed.

https://github.com/apache/accumulo/runs/1078818587

It happened with commit f4e3c51efed22f9c17185bf023883ecb1962e6e9 . It might just have been a transient issue with GitHub Actions containers, or an unrelated problem. It did not seem to reoccur with your latest commits.

Speaking of your commits. You don't need to force-push when update the PR. In fact, it actually makes it harder to review, because it GitHub sends notifications to followers showing the diff between the current code and the previous... but that link doesn't work on force-pushes, because the earlier commits are no longer present. When it comes time to merge, we'll squash it all down to one, anyway, so it's better to just add commits instead of force-pushing.

**Review by ctubbsii — APPROVED** [review]
Good test. It might be good practice to verify the mock objects using EasyMock.verify at the end of the tests. However, you already know the methods were called, because you're checking the returned value from the method, so it's probably not necessary. I'll hold off on merging today, in case you want to add that, to strengthen the test.

Thanks for the PR!

**Comment by jkosh44** [comment]
@ctubbsii Sorry about the force pushing, bad habits. I'll add the verify to my tests, thanks for the info, I'm not as familiar with EasyMock as I'd like to be.

Did you see my comment about the prefix subset?

**Comment by ctubbsii** [comment]
> Did you see my comment about the prefix subset?

I missed it the first time. I think I see what you're suggesting, but I don't think it's worth it. If you can show that it's a substantive performance advantage, then it might make sense... but I suspect it'd be a negligible difference, and just more code complexity to maintain, without real benefit.

File: server/base/src/test/java/org/apache/accumulo/server/ServiceEnvironmentImplTest.java
Comment (ctubbsii) [inline server/base/src/test/java/org/apache/accumulo/server/ServiceEnvironmentImplTest.java:54]: The verify method is actually a varargs, so you can pass both in one:

```suggestion
  @After
  public void verifyMocks() {
    verify(srvCtx, acfg);
  }

```

Comment (jkosh44) [reply server/base/src/test/java/org/apache/accumulo/server/ServiceEnvironmentImplTest.java]: Fixed

**Comment by jkosh44** [comment]
I agree with the complexity comment. I'm good to merge if you are.

**Comment by ctubbsii** [comment]
Thanks for the PR @jkosh44 !

I see you've contributed before, but you're not on our "people" page, listing contributors. If you wish to be added as a contributor to https://accumulo.apache.org/people/ , please open a pull request to add yourself at https://github.com/apache/accumulo-website/edit/master/pages/people.md and leave a reference to `apache/accumulo#1701` in your commit log.

If you intend to be a regular contributor to Accumulo projects, please consider subscribing to our developer mailing list (https://accumulo.apache.org/contact-us/) and introducing yourself. :smiley_cat:

**Comment by jkosh44** [comment]
Thanks for the review @ctubbsii! 
I'm not sure how regularly I plan on contributing, but I'll do all that now.

**Pull request merged**
